sigs/jvm/rfcs/20200126-tensorflow-data.md

```diff
@@ -0,0 +1,209 @@
+transformed. Each element is a list of "output" operands, represented by the type `List<Output<?>>`. 
+
+Note: An `Output` is a symbolic handle to a tensor produced by a TensorFlow op. In graph
+mode, `Output` objects will not have a concrete `Tensor` value unless all dependent operations
+are run in a `Session` (this is done "in-real-time" in eager mode).
+
+### Construction
+
+Datasets can be constructed either directly from a data source (e.g. a list of tensors representing the components of the dataset), or as a transformation on an existing dataset.
+
+#### From Data Source
+
+To construct a dataset from a list of tensor components, use 
+`Dataset.fromTensorSlices( ... )`. For example, say we are working
+with a standard feature/label dataset which has 4 elements.
+
+```java
+float[][] features = new float[][] {
+    {1, 2, 3},
+    {4, 5, 6},
+    {7, 8, 9},
+    {10, 11, 12}
+}
+
+float[] labels = new float[] {
+    0,
+    1,
+    1,
+    0
+}
+```
+
+A dataset can be constructed from a list of the constant `Operand`s generated
+from this dataset, and a list of `DataType` objects corresponding
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
+    Arrays.asList(TInt32.DTYPE, TInt32.DTYPE)
+);
+```
+
+
+Other data sources are also possible, using `tf.data` ops. Datasets can also be constructed from 
+
+#### Transformations on Existing Datasets
+
+Once a dataset has been created from a data source it can be transformed by calling
+methods on the `Dataset` object. For example, to group elements in the above dataset into batches of size `2`, use `Dataset.batch(int batchSize)`:
+
+```java
+dataset = dataset.batch(2)
+```
+
+Dataset transformations alter both the values and shapes of the original elements, and
+return a *new* `Dataset` object.
+
+In this case, the original dataset had 4 elements of shape `[features: (3,) labels: (1,)]`.
+Once the `.batch` transformation is applied, the new dataset has 2 elements (batches) of shape `[features: (2, 3), labels: (2, 1)]`.
+
+Similar transformations include `.skip`, `.take`, `.map`, `.filter`, etc.
+
+
+### Iterating over Dataset Elements
+
+The primary use of a dataset is for iteration over its elements.
+Each row (or batch) element is represented as a list of tensor components, with
+type `List<Output<?>>`. The tensor components of this elements can be accessed using `List.get(int index)`.
+
+It is recommended to use `Tensor.expect(DataType<?> dtype)` to restore types
+to the retrieved tensors.
+
+
+#### Eager Mode: Iterable
+The `Dataset` class implements the `Iterable` interface, so in
+eager mode, iteration over dataset elements is possible using a standard for-each loop.
```

Comment:
that could still be useful in graph mode, the only difference is that you won't be able to access the tensor or operate directly on the outputs in your iteration. But you could do other stuff with the operands.

Reply:
@Hmm; wondering what the semantics of this would be. In graph mode, you don't end up getting a new `List<Output<?>>` for each "batch"; you use one `List<Output<?>>`.

so allowing:

```
for (List<Output<?>> : graphDataset);
```
would be a bit strange?

Would this be more like a `map` that operates over individual elements? We can allow

```
graphDataset.map(Function<List<Output<?>>, List<Output<?>>);
```

which would probably accomplish the same thing?

Reply:
I think it's best to keep the `while` loop, as it's consistent with TF Python's API, and other semantics start to get really strange, since it would hide a `session.run` call.